seastar/src/file.rs

```
@@ -0,0 +1,407 @@
+        file.close().await.unwrap();
+        assert_eq!(res.0, msg.len());
+        let bytes = &res.1.as_slice()[..msg.len()];
+        assert_eq!(bytes, msg);
+    }
+
+    #[seastar::test]
+    async fn test_file_read_dma_big() {
+        let msg = (0..CHUNK_SIZE * 5)
+            .map(|_| rand::random::<u8>())
+            .collect::<Vec<u8>>();
+        std::fs::OpenOptions::new()
+            .create(true)
+            .write(true)
+            .open("example-read-big")
+            .unwrap()
+            .write_all(msg.as_ref())
+            .unwrap();
+        let file = OpenOptions::new()
+            .read(true)
+            .open("example-read-big")
+            .await
+            .unwrap();
+        let buffer = DmaBuffer::from_vec(vec![0u8; CHUNK_SIZE * 5]);
+        let res = file.read_dma(buffer, 0).await.unwrap();
+        file.close().await.unwrap();
+        assert_eq!(res.0, msg.len());
+        let bytes = res.1.as_slice();
+        assert_eq!(bytes, msg.as_slice());
+    }
+
+    #[seastar::test]
+    async fn test_file_write_dma() {
+        let v = (0..CHUNK_SIZE).map(|_| rand::random::<u8>()).collect();
+        let buffer = DmaBuffer::from_vec(v);
+        let file = OpenOptions::new()
+            .create(true)
+            .write(true)
+            .open("example-write")
+            .await
+            .unwrap();
+        let res = file.write_dma(buffer, 0).await.unwrap();
+        file.flush().await.unwrap();
+        file.close().await.unwrap();
+        assert_eq!(res.0, CHUNK_SIZE);
+        let mut line = Vec::new();
+        std::fs::OpenOptions::new()
+            .read(true)
+            .open("example-write")
+            .unwrap()
+            .read_to_end(&mut line)
+            .unwrap();
+        let bytes = res.1.as_slice();
+        assert_eq!(bytes, line.as_slice());
+    }
+
+    #[seastar::test]
+    async fn test_file_write_dma_big() {
+        let v = (0..10000).map(|_| rand::random::<u8>()).collect();
```

The size of `v` is not aligned to 4 kilobytes. `DmaBuffer::new()` will truncate some of the bytes at the end and will only write 8 kilobytes to disk.

---

Besides, I don't think `from_vec` should truncate. I would prefer if it panicked or returned an error if the vec's length is not a multiple of 4kb. Silently dropping data like this really not cool.

---

It was intentional to make a such a test. I thought that `DmaBuffer` should truncate excessive bytes or fill a chunk with `0u8` if vec is too small. But having 4kB alignment assert sounds better.